### DISCUSSION TITLE
when_activated event not firing when is_active is set for pull down conditions

### DISCUSSION THREAD

**Discussion by sreekarkrishna** [gpiozero/gpiozero#1035] [Q&A]
I have a scenario where I am interfacing a Pi 3B to detect activation of a sensor. The sensor by default only offers a rising edge detection scenario. I have setup a `gpiozero.Button` on GPIO 17, with `pull_up=False`. 

I am facing this strange situation where the `button.is_active` is correctly set when the sensor is active, but the `button.when_activated` event is not firing on the rising edge. When the sensor deactivates, both the `when_activated` and `when_deactivated` events are fired. 

I wrote a test script which sets up the event handlers, and also sets up a while loop to monitor the `button.is_active `property. 

```
from gpiozero import Button
from time import sleep
from gpiozero import DigitalInputDevice
from datetime import datetime

//# event handlers
def activated(self):
    print(f"{datetime.now()}: Activated {str(self.pin.number)}")

def deactivated(self):
    print(f"{datetime.now()}: Deactivated {str(self.pin.number)}")

//# Set GPIO 17 with pull down resistor
button = Button(17, pull_up=False,  bounce_time=0.2)

//# Set up handlers
button.when_activated = activated
button.when_deactivated = deactivated

// Loop to monitor button active state
state = button.is_active
while (True):

    //# Create a debounce
    condition1 = button.is_active != state
    if condition1:
        sleep(0.4)
        new_state = button.is_active
        condition2 = new_state != state
        if condition2:

            //# Print states and set the stored state to new state 
            print(f"{datetime.now()}: Stored state value is {state} and Button is {new_state}")
            state = new_state
```

Here is the output of the code:
```
2022-03-12 07:52:11.679051: Stored state value is False and Button is True
2022-03-12 07:52:20.950999: Activated 17
2022-03-12 07:52:21.146104: Deactivated 17
2022-03-12 07:52:21.351971: Stored state value is True and Button is False
```

The 'button.is_active' was detected in the while loop at 7:52:11, but the `when_activated `event fired only when the sensor deactivated and the two events (`when_activated`, `when_deactivated`) are packed along with the 'button.is_active' detection on the falling edge.

Any thoughts what might be going on?

**Comment by lurch** [comment]
Could you wrap your code above in triple-backticks please, so that the indentation is retained?

> Any thoughts what might be going on?

Might be a bug in the EventsMixin? Does it _continually_ fire both activated and deactivated at the same time, or does it only happen after the first event and then settle down afterwards?
Does it behave any differently if you use DigitalInputDevice rather than Button? Why have you got both `bounce_time=` and also `My own debounce` ?

**Reply by sreekarkrishna** [reply]
Thanks for the quick response. 

Yes, it continually fires both activated and deactivated at the same time. 

DigitalInputDevice is showing the same behavior. 

I used the while loop to check if the is_active state changing correctly. The sensor generates a lot of noise, hence I built my own debounce for cross verification. In my ideal world, the handlers work and I will be packaging this code as part of a Python TCP server.

Random Hypothesis: I am reading through the code base to understand how the ThreadQueue works, and it there is a specific reason the queue is emptying out at all at once on the falling edge trigger.

**Reply by lurch** [reply]
Could you try replicating this with an _actual_ button, just to rule out your "very noisy sensor" as being part of the problem?

And could you check your code above please? You've got an invalid `def changed()` line, and you're creating the `button` object twice?

**Reply by sreekarkrishna** [reply]
Sorry, I must have copied and pasted code that was part of my testing. I have fixed it. 

I tested the code with a physical button, and it worked as expected (according to the gpiozero documentation), but that still doesnt answer two questions:

1. Why did the debounce in the `DigitalInputDevice` not take care of the 'noisy sensor'? I am able to take care of the noisy sensor in my simple while loop. I would have expected a similar functionality with the higher order class like `DigitalInputDevice`, or `Button`. 
2. Why are `when_activated` and `when_deactivated` being called exactly once, but executed at the falling edge of the sensor? I would have expected the noisy sensor to create a series of handler triggers, but that is not the case, and I have exactly one of each handler.

Finally, on the same RPi, I also have other sensors (different type of sensor) which has a falling edge (active-low) output. I run them through the different set of GPIO pins, and the `Button` class handles it beautifully. I fell into this dark hole of debugging because the operation of gpiozero for a `pull_up` resistor is very different than with a `pull_down` resistor.

**Reply by lurch** [reply]
The internals of gpiozero are quite complicated, so I suspect those are all questions that only @waveform80 can answer, and unfortunately he seems to be very busy with other projects.
Hmmm it might be related to https://github.com/gpiozero/gpiozero/discussions/937  ? :man_shrugging: 

Could you add a capacitor to your noisy sensor to "debounce" in hardware? It _might_ make gpiozero more reliable?

EDIT: i.e. I **think** that the problem is "just" your noisy sensor, rather than "the operation of gpiozero for a pull_up resistor is very different than with a pull_down resistor.". Sorry I can't be more helpful.

**Reply by sreekarkrishna** [reply]
Thanks lurch for taking a look at this issue. I agree that gpiozero is very complicated code base. I am trying to run the complete code base on my RPi and see if I can debug it through an IDE with breakpoints and logging setup. 

I just tested the system again using the blocking handlers `Button.wait_for_active `and `Button.wait_for_inactive`.  The same issue continues. 

```
from gpiozero import Button

button = Button(17, pull_up=False, bounce_time=0.2)

button.wait_for_active()
print(f"{datatime.now()}: Activated")

button.wait_for_inactive()
print(f"{datatime.now()}: Deactivated")
```

This is the simplest code that is not working according to plan. My code stays blocked until the sensor goes active, and then inactive, and finally both print statements come out. Again, no noisy multiple activations, or deactivations, it does exactly one activation and one deactivation, but both of the falling edge of the sensor output.
